grove/deutsch_jozsa/deutsch_jozsa.py

```diff
@@ -0,0 +1,159 @@
+        :return: None
+        :rtype: NoneType
+        """
+        self.bit_map = bitstring_map
+        self.n_qubits = len(list(bitstring_map.keys())[0])
+        # We use one extra qubit for making the oracle,
+        # and one for storing the answer of the oracle.
+        self.n_ancillas = 2
+        self._qubits = list(range(self.n_qubits + self.n_ancillas))
+        self.computational_qubits = self._qubits[:self.n_qubits]
+        self.ancillas = self._qubits[self.n_qubits:]
+        self.unitary_matrix = self.unitary_function(bitstring_map)
+        self.deutsch_jozsa_circuit = self._construct_deutsch_jozsa_circuit()
+
+    def _construct_deutsch_jozsa_circuit(self):
+        """
+        Builds the Deutsch-Jozsa circuit. Which can determine whether a function f mapping
+        :math:`\{0,1\}^n \to \{0,1\}` is constant or balanced, provided that it is one of them.
+
+        :return: A program corresponding to the desired instance of Simon's Algorithm.
+        :rtype: Program
+        """
+        dj_prog = pq.Program()
+
+        # Put the first ancilla qubit (query qubit) into minus state
+        dj_prog.inst(X(self.ancillas[0]), H(self.ancillas[0]))
+
+        # Apply Hadamard, Oracle, and Hadamard again
+
+        dj_prog.inst([H(qubit) for qubit in self.computational_qubits])
+
+        # Build the oracle
+        oracle_prog = pq.Program()
+        oracle_prog.defgate(ORACLE_GATE_NAME, self.unitary_matrix)
+
+        scratch_bit = self.ancillas[1]
+        qubits_for_funct = [scratch_bit] + self.computational_qubits
+        oracle_prog.inst(tuple([ORACLE_GATE_NAME] + qubits_for_funct))
+        dj_prog += oracle_prog
+
+        # Here the oracle does not leave the computational qubits unchanged, so we use a CNOT to
+        # to move the result to the query qubit, and then we uncompute with the dagger.
+        dj_prog.inst(CNOT(self._qubits[0], self.ancillas[0]))
+        dj_prog += oracle_prog.dagger()
+        dj_prog.inst([H(qubit) for qubit in self.computational_qubits])
+        return dj_prog
+
+    @staticmethod
+    def unitary_function(mappings):
+        """
+        Creates a unitary transformation that maps each state to the values specified
+        in mappings.
+
+        Some (but not all) of these transformations involve a scratch qubit, so room for one is
+        always provided. That is, if given the mapping of n qubits, the calculated transformation
+        will be on n + 1 qubits, where the 0th is the scratch bit and the return value
+        of the function is left in the 1st.
+
+        :param mappings: Dictionary of the mappings of f(x) on all length n bitstrings, e.g.
+
+            >>> {'00': '0', '01': '1', '10': '1', '11': '0'}
+
+        :type mappings: Dict[String, Int]
+        :return: ndarray representing specified unitary transformation.
+        :rtype: np.ndarray
+        """
+        num_qubits = int(np.log2(len(mappings)))
+        bitsum = sum([int(bit) for bit in mappings.values()])
```

Comment:
this is interesting! Now we need a loop over the full function! This is O(2**n) which is the classic system's runtime. Are you implying we cannot construct the oracle faster than the classic algorithm runtime? If so then what's the point of DJ? We should highlight this!

Reply:
So this implementation (written over the summer) constructs an oracle for a given function. What part specifically are you saying is 2**n? Certainly we're `defgate`- ing the oracle, and that matrix is exponentially large.

Normally, in any of these cases, you're handed an oracle. We're explicitly defining example oracles, and so to do that in full generality it seems to be that this is going to require this kind of scaling, right?

Reply:
yes, the def-gate is exponentially large, but it's also super-sparse. do the non-sparse elements also scale exponentially in their generation? If yes, how do other problems get around this?

Reply:
I'l address this in documentation.